sandbox/lua/modules/cbuf_utils.lua

```diff
@@ -0,0 +1,151 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[[
+API
+---
+**parse_cbuf(cbuf_text)**
+
+    Parses the circular buffer data such as that output by a SandboxFilter and
+    returns a table of headers and an array containing the rows of data.
+
+    *Arguments*
+        - cbuf_text (string)
+            String containing the textual circular buffer data. Usually the
+            entire payload of a cbuf message generated by a SandboxFilter.
+
+    *Return*
+        - If parsing is successful, two values are returned:
+          - headers:
+            A table containing the header values of `time`, `rows`,
+            `seconds_per_row`, `columns`, and `column_info`.
+          - rows:
+            An array or nested row arrays, where each row array contains the
+            column values, in order, either as a number or as the string
+            literal "nan". Note that the last row should typically not be
+            used, as it is probably incomplete.
+        - If parsing fails, or the header data isn't valid, nil is returned.
+
+**get_start_idx(last_time, headers)**
+
+    Given the timestamp of a previous cbuf payload and the headers from a more
+    current cbuf payload, calculates the index of the first row of new data in
+    the more current cbuf data set.
+
+    *Arguments*
+        - last_time (number)
+            The `time` value retrieved from the headers table of the earlier
+            cbuf data.
+        - headers (table)
+            The headers table (as returned from `parse_cbuf`) of the current
+            cbuf data.
+
+    *Return*
+
+        - If calculations are successful, will return a positive integer
+          number value corresponding to the index of the first row that
+          contains new data since the previous cbuf time.
+        - If the new cbuf time is the same as the provided last_time, will
+          return -2.
+        - If calculations fail (because last_time is more recent than new
+          time, or because the time interval isn't exactly divisible by the
+          cbuf's seconds_per_row value), will return -1.
+--]]
+
+local cjson = require "cjson"
+local string = require "string"
+local pcall = pcall
+local tonumber = tonumber
+
+local M = {}
+setfenv(1, M) -- Remove external access to contain everything in the module.
+
+local function check_headers(headers)
+    if (not headers.time) or (not headers.rows) or (not headers.seconds_per_row)
+        or (not headers.columns) or (not headers.column_info) then
+
+        return
+    end
+    if headers.columns ~= #headers.column_info then
+        return
+    end
+    return true
+end
+
+-- [[ Public Interface --]]
+
+function parse_cbuf(cbuf_text)
```

Comment:
See https://github.com/mozilla-services/lua_sandbox/blob/dev/modules/cbufd.lua#L47 for a similar parser